### PR TITLE
[CLOUD-163] searchcontexts: only org member can search with org namespace on Cloud

### PR DESCRIPTION
**Pull request by unknwon**
Limiting using search contexts under an organization namespace to only members of the organization on Sourcegraph Cloud.

_It is recommended to review [without whitespaces](https://github.com/sourcegraph/sourcegraph/pull/29091/files?w=1)._

Below is what it looks like in the UI:
CleanShot 2021-12-16 at 14 54 29@2x

---

Jira: CLOUD-163

### CODE REVIEW

File: internal/search/searchcontexts/search_contexts_test.go [hunk 1]
```diff
@@ -59,17 +60,53 @@ func TestResolvingValidSearchContextSpecs(t *testing.T) {
...
+	orig := envvar.SourcegraphDotComMode()
+	envvar.MockSourcegraphDotComMode(true)
+	defer envvar.MockSourcegraphDotComMode(orig)
+
+	tests := []struct {
+		name                  string
+		searchContextSpec     string
+		wantSearchContextName string
+	}{
+		{name: "resolve organization search context", searchContextSpec: "@org", wantSearchContextName: "org"},
```
Comment (unknwon) [inline internal/search/searchcontexts/search_contexts_test.go:82]: The entire test body is copied from `TestResolvingValidSearchContextSpecs`, there is only one test but what's bad about it being table-driven-ready? 😄

File: internal/search/searchcontexts/search_contexts_test.go [hunk 2]
```diff
@@ -99,19 +141,55 @@ func TestResolvingInvalidSearchContextSpecs(t *testing.T) {
...
+	orig := envvar.SourcegraphDotComMode()
+	envvar.MockSourcegraphDotComMode(true)
+	defer envvar.MockSourcegraphDotComMode(orig)
+
+	tests := []struct {
+		name              string
+		searchContextSpec string
+		wantErr           string
+	}{
+		{name: "org not a member", searchContextSpec: "@org-not-member", wantErr: "namespace not found"},
```
Comment (unknwon) [inline internal/search/searchcontexts/search_contexts_test.go:163]: The entire test body is copied from `TestResolvingInvalidSearchContextSpecs`, there is only one test but what's bad about it being table-driven-ready? 😄

File: internal/search/searchcontexts/search_contexts_test.go [hunk 3]
```diff
@@ -87,7 +124,12 @@ func TestResolvingInvalidSearchContextSpecs(t *testing.T) {
 	}
 
 	ns := dbmock.NewMockNamespaceStore()
-	ns.GetByNameFunc.SetDefaultReturn(&database.Namespace{}, nil)
+	ns.GetByNameFunc.SetDefaultHook(func(ctx context.Context, name string) (*database.Namespace, error) {
+		if name == "org-not-member" {
```
Comment (novoselrok) [inline internal/search/searchcontexts/search_contexts_test.go:128]: Why is this change necessary? I don't think `org-not-member` is used in this test?

File: internal/search/searchcontexts/search_contexts_test.go (hunk 2 above)
Comment (novoselrok) [inline internal/search/searchcontexts/search_contexts_test.go:163]: If I understand the changes correctly, then a non-org member cannot access `@org-not-member/random`, right? Could you add this test case as well?

**Review by novoselrok — APPROVED** [review]

File: internal/search/searchcontexts/search_contexts_test.go (hunk 3 above)
Comment (unknwon) [reply internal/search/searchcontexts/search_contexts_test.go:128]: Good catch!... Missed this when I separate tests into two.

File: internal/search/searchcontexts/search_contexts_test.go (hunk 2 above)
Comment (unknwon) [reply internal/search/searchcontexts/search_contexts_test.go:163]: Sure thing!

File: internal/search/searchcontexts/search_contexts.go [hunk 4]
```diff
@@ -54,21 +56,40 @@ func ResolveSearchContextSpec(ctx context.Context, db database.DB, searchContext
...
-		})
-	} else if hasNamespaceName && !hasSearchContextName {
-		namespace, err := db.Namespaces().GetByName(ctx, parsedSearchContextSpec.NamespaceName)
-		if err != nil {
-			return nil, err
+
+		// Only member of the organization can use search contexts under the
+		// organization namespace on Sourcegraph Cloud.
+		if envvar.SourcegraphDotComMode() && namespace.Organization > 0 {
+			_, err = db.OrgMembers().GetByOrgIDAndUserID(ctx, namespace.Organization, actor.FromContext(ctx).UID)
```
Comment (rafax) [inline internal/search/searchcontexts/search_contexts.go:70]: Is there a reason why you can't use backend.CheckOrgAccess here? Dependency problem? Performance?

**Review by rafax — APPROVED** [review]

File: internal/search/searchcontexts/search_contexts.go (hunk 4 above)
Comment (unknwon) [reply internal/search/searchcontexts/search_contexts.go:70]: I thought about it but `backend` is a subpackage of `frontend`, and this `searchcontexts` seems have well-abstracted and may introduce circular dependency by importing `backend` package.
